Javadoc Mojo cannot handle true "modular" projects

For projects using the "module path" instead of the "class path" the javadoc mojo does not work. You see errors such as this:

```
[INFO] --- xmvn-mojo:3.1.0:javadoc (default-cli) @ jakarta.xml.bind-api ---
/home/mbooth/git/jaxb-api/jaxb-api/src/main/java/module-info.java:18: error: module not found: jakarta.activation
    requires transitive jakarta.activation;
                               ^
```

I was able to successfully build Javadoc with xmvn by making this change:

```
diff --git a/xmvn-mojo/src/main/java/org/fedoraproject/xmvn/mojo/JavadocMojo.java b/xmvn-mojo/src/main/java/org/fedoraproject/xmvn/mojo/JavadocMojo.java
index 173bdaa9..e140852d 100644
--- a/xmvn-mojo/src/main/java/org/fedoraproject/xmvn/mojo/JavadocMojo.java
+++ b/xmvn-mojo/src/main/java/org/fedoraproject/xmvn/mojo/JavadocMojo.java
@@ -215,7 +215,7 @@ public class JavadocMojo
             opts.add( "-version" );
             opts.add( "-Xdoclint:none" );
 
-            opts.add( "-classpath" );
+            opts.add( "--module-path" );
             opts.add( quoted( StringUtils.join( getClasspath().iterator(), ":" ) ) );
             opts.add( "-encoding" );
             opts.add( quoted( encoding ) );
```

But I would be surprised if it was as simple as passing `--module-path` instead of `-classpath` for modular projects. The maven-javadoc-plugin seems to have some amount of code for dealing with it: https://github.com/apache/maven-javadoc-plugin/blob/master/src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java

--- a/xmvn-it/src/test/java/org/fedoraproject/xmvn/it/maven/mojo/JavadocIntegrationTest.java
+++ b/xmvn-it/src/test/java/org/fedoraproject/xmvn/it/maven/mojo/JavadocIntegrationTest.java
@@ -41,4 +41,16 @@ public class JavadocIntegrationTest
         assertTrue( Files.isSymbolicLink( Paths.get( ".xmvn/apidocs" ) ) );
         assertTrue( Files.isSameFile( Paths.get( ".xmvn/apidocs" ), Paths.get( "target/xmvn-apidocs" ) ) );
     }
+
+    @Test
+    public void testJavadocJPMS()
+        throws Exception
+    {
+        performTest( "verify", "org.fedoraproject.xmvn:xmvn-mojo:javadoc" );
+
+        assertTrue( Files.isDirectory( Paths.get( "target/xmvn-apidocs" ) ) );
+        assertTrue( Files.isRegularFile( Paths.get( "target/xmvn-apidocs/test.jpms/foo/Bar.html" ) ) );
+        assertTrue( Files.isSymbolicLink( Paths.get( ".xmvn/apidocs" ) ) );
+        assertTrue( Files.isSameFile( Paths.get( ".xmvn/apidocs" ), Paths.get( "target/xmvn-apidocs" ) ) );
+    }
 }

--- a/xmvn-it/src/test/resources/testJavadocJPMS/.xmvn/configuration.xml
+++ b/xmvn-it/src/test/resources/testJavadocJPMS/.xmvn/configuration.xml
@@ -1,0 +1,7 @@
+<configuration>
+  <resolverSettings>
+    <metadataRepositories>
+      <repository>metadata.xml</repository>
+    </metadataRepositories>
+  </resolverSettings>
+</configuration>

--- a/xmvn-it/src/test/resources/testJavadocJPMS/pom.xml
+++ b/xmvn-it/src/test/resources/testJavadocJPMS/pom.xml
@@ -1,0 +1,24 @@
+<project>
+  <modelVersion>4.0.0</modelVersion>
+  <groupId>xmvn.its</groupId>
+  <artifactId>javadoc-test-jpms</artifactId>
+  <version>0.0.0-SNAPSHOT</version>
+  <dependencies>
+    <dependency>
+      <groupId>org.junit.jupiter</groupId>
+      <artifactId>junit-jupiter-api</artifactId>
+      <version>5.6.0</version>
+    </dependency>
+  </dependencies>
+  <build>
+    <plugins>
+      <plugin>
+        <artifactId>maven-compiler-plugin</artifactId>
+        <version>3.8.1</version>
+        <configuration>
+          <release>11</release>
+        </configuration>
+      </plugin>
+    </plugins>
+  </build>
+</project>

--- a/xmvn-it/src/test/resources/testJavadocJPMS/src/main/java/foo/Bar.java
+++ b/xmvn-it/src/test/resources/testJavadocJPMS/src/main/java/foo/Bar.java
@@ -1,0 +1,24 @@
+/*-
+ * Copyright (c) 2020 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package foo;
+
+import org.junit.jupiter.api.TestReporter;
+
+public class Bar implements TestReporter
+{
+	public void publishEntry(java.util.Map<String,String> map) {
+	}
+}

--- a/xmvn-it/src/test/resources/testJavadocJPMS/src/main/java/module-info.java
+++ b/xmvn-it/src/test/resources/testJavadocJPMS/src/main/java/module-info.java
@@ -1,0 +1,3 @@
+module test.jpms {
+	requires transitive org.junit.jupiter.api;
+}

--- a/xmvn-mojo/src/main/java/org/fedoraproject/xmvn/mojo/JavadocMojo.java
+++ b/xmvn-mojo/src/main/java/org/fedoraproject/xmvn/mojo/JavadocMojo.java
@@ -198,6 +198,18 @@ public class JavadocMojo
     public void execute()
         throws MojoExecutionException, MojoFailureException
     {
+        Path javadocExecutable;
+        if ( System.getenv().containsKey( "JAVA_HOME" ) )
+        {
+            javadocExecutable = Paths.get( System.getenv( "JAVA_HOME" ) ) //
+                                     .resolve( "bin" ) //
+                                     .resolve( "javadoc" );
+        }
+        else
+        {
+            javadocExecutable = Paths.get( "/usr/bin/javadoc" );
+        }
+
         try
         {
             if ( StringUtils.isEmpty( encoding ) )
@@ -242,7 +254,7 @@ public class JavadocMojo
             opts.add( "-version" );
             opts.add( "-Xdoclint:none" );
 
-            opts.add( "-classpath" );
+            opts.add( isJavadocModular( javadocExecutable ) ? "--module-path" : "-classpath" );
             opts.add( quoted( StringUtils.join( getClasspath().iterator(), ":" ) ) );
             opts.add( "-encoding" );
             opts.add( quoted( encoding ) );
@@ -266,18 +278,6 @@ public class JavadocMojo
                 opts.add( quoted( file ) );
 
             Files.write( outputDir.resolve( "args" ), opts, StandardOpenOption.CREATE );
-
-            Path javadocExecutable;
-            if ( System.getenv().containsKey( "JAVA_HOME" ) )
-            {
-                javadocExecutable = Paths.get( System.getenv( "JAVA_HOME" ) ) //
-                                         .resolve( "bin" ) //
-                                         .resolve( "javadoc" );
-            }
-            else
-            {
-                javadocExecutable = Paths.get( "/usr/bin/javadoc" );
-            }
 
             ProcessBuilder pb = new ProcessBuilder( javadocExecutable.toRealPath().toString(), "@args" );
             pb.directory( outputDir.toFile() );

--- a/xmvn-mojo/src/main/java/org/fedoraproject/xmvn/mojo/JavadocMojo.java
+++ b/xmvn-mojo/src/main/java/org/fedoraproject/xmvn/mojo/JavadocMojo.java
@@ -15,8 +15,11 @@
  */
 package org.fedoraproject.xmvn.mojo;
 
+import java.io.BufferedReader;
 import java.io.File;
 import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
 import java.lang.ProcessBuilder.Redirect;
 import java.nio.file.DirectoryStream;
 import java.nio.file.Files;
@@ -93,6 +96,30 @@ public class JavadocMojo
         arg = StringUtils.replace( arg, "\\", "\\\\" );
         arg = StringUtils.replace( arg, "'", "\\'" );
         return "'" + arg + "'";
+    }
+
+    static boolean isJavadocModular( Path javadocExecutable )
+        throws IOException, InterruptedException, MojoExecutionException
+    {
+        ProcessBuilder pb = new ProcessBuilder( javadocExecutable.toRealPath().toString(), "-help" );
+        pb.redirectInput( new File( "/dev/null" ) );
+        pb.redirectError( Redirect.INHERIT );
+        Process process = pb.start();
+
+        try ( InputStream is = process.getInputStream();
+                        InputStreamReader isr = new InputStreamReader( is );
+                        BufferedReader br = new BufferedReader( isr ) )
+        {
+            return br.lines().anyMatch( line -> line.contains( "module-path" ) );
+        }
+        finally
+        {
+            int exitCode = process.waitFor();
+            if ( exitCode != 0 )
+            {
+                throw new MojoExecutionException( "Javadoc failed with exit code " + exitCode );
+            }
+        }
     }
 
     private static void findJavaSources( Collection<Path> javaFiles, Path dir )

--- a/xmvn-mojo/src/test/java/org/fedoraproject/xmvn/mojo/JavadocMojoTest.java
+++ b/xmvn-mojo/src/test/java/org/fedoraproject/xmvn/mojo/JavadocMojoTest.java
@@ -1,0 +1,71 @@
+/*-
+ * Copyright (c) 2014-2020 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.fedoraproject.xmvn.mojo;
+
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assertions.fail;
+import static org.junit.jupiter.api.Assumptions.assumeTrue;
+
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+import org.apache.maven.plugin.MojoExecutionException;
+import org.junit.jupiter.api.Test;
+
+/**
+ * Unit tests for javadoc MOJO.
+ * 
+ * @author Mikolaj Izdebski
+ */
+public class JavadocMojoTest
+{
+    @Test
+    public void testJavadocIsModular8()
+        throws Exception
+    {
+        Path javadocPath = Paths.get( "/usr/lib/jvm/java-1.8.0/bin/javadoc" );
+        assumeTrue( Files.isExecutable( javadocPath ) );
+        assertFalse( JavadocMojo.isJavadocModular( javadocPath ) );
+    }
+
+    @Test
+    public void testJavadocIsModular11()
+        throws Exception
+    {
+        Path javadocPath = Paths.get( "/usr/lib/jvm/java-11/bin/javadoc" );
+        assumeTrue( Files.isExecutable( javadocPath ) );
+        assertTrue( JavadocMojo.isJavadocModular( javadocPath ) );
+    }
+
+    @Test
+    public void testJavadocIsModularError()
+        throws Exception
+    {
+        Path javadocPath = Paths.get( "/bin/false" );
+        assumeTrue( Files.isExecutable( javadocPath ) );
+        try
+        {
+            JavadocMojo.isJavadocModular( javadocPath );
+            fail();
+        }
+        catch ( MojoExecutionException e )
+        {
+            assertTrue( e.getMessage().startsWith( "Javadoc failed" ) );
+        }
+    }
+}

